### DISCUSSION TITLE
Under Ubuntu22.04 `hotplugtest` result in `No access to device: Access denied (insufficient permissions)`

### DISCUSSION THREAD

**Discussion by brmarkus** [libusb/libusb#1515] [Q&A]
Coming from "https://libusb.info/", which states:

> It is **user-mode**: No special privilege or elevation is required for the application to communicate with a device.

However, when running `hotplugtest` and insert/eject a USB storage (simple USB mass storage pen drive, 32GB) I get the following:

$> ./hotplugtest 
Device detached: 0951:1621
Device attached: 0951:1621
No access to device: Access denied (insufficient permissions)

Is it expected to require root privileges (under Linux) to request device details like Manufacturer, Vendor, Product, SerialNumber?

Running under Ubuntu 22.04LTS.
Kernel 6.5.0-35
libusb from "libusb-1.0.27.tar.bz2"

**Comment by tormodvolden** [comment]
Yes, the system can protect devices against non-privileged user access. It is the same as you cannot usually do "echo hello > /dev/sda1". So if you don't have write access to the device node /dev/bus/usb/xxx/yyy you cannot send control transfers to request device details. The kernel has however already retrieved and cached some of this information, so you can find them under /sys/bus/usb/devices . There are some ideas about enhancing libusb to report these values without the user having to open the device, see #1258.

**Reply by brmarkus** [reply]
I had hoped that the underlying layers would make a difference between read-only (like querying device details) and write-requests; in my case I want to get as many device attributes as possible (without changing anything, not mounting the device, no handshaking with the device).
Let's see what I can make with the values under the mentioned `/sys/bus/usb/devices`.
Thank you for the reference to #1258 (and its other references) - will subscribe!

**Comment by tormodvolden** [comment]
You can also take a look at how lsusb uses libudev to collect information from the system HW database. https://github.com/gregkh/usbutils/blob/master/names.c

**Reply by brmarkus** [reply]
Yeah, already looked into that.
Because I was wondering why `lsusb` and `libusb` are printing different strings:

libusb: `Manufacturer:              Linux 6.5.0-35-generic xhci-hcd`
lsusb and libudev: `Linux Foundation`

and found e.g. "/usr/share/misc/usb.ids" where the IDs could be looked-up with.
